blog/2022-11-03-short-circuiting.md

```diff
@@ -0,0 +1,162 @@
+---
+title: Array short-circuiting
+author: Philip Munksgaard
+description: In-place updates without overhead
+---
+
+Hi, Philip Munksgaard here! I'm a PhD student working on Futhark, and I'm here
+to talk about some of the stuff I've been working on for the past couple of
+years. Hope you enjoy it!
+
+Futhark is a parallel high-level array-oriented language.  Parallelism is explicitly
+expressed by the user in terms of the second-order array combinators such as
+`map`, `reduce` and `scan`.  Because parallelism is explicitly described, the
+compiler is able to make certain guarantees about the parallelism, such as the
+guarantee that there are no data races in the code.
+
+To see why this matters, consider this simple stencil function:
+
+```
+def stencil [n] (xs: *[n]i64): [n]i64 =
+  xs with [1:n-1] = map (\i -> xs[i-1] + xs[i] + xs[i+1]) (1..<n-1)
+```
+
+In imperative languages, you might try implement this in-place, but that would
+cause data-dependency errors.  Specifically, it is possible for one thread to
+update an element of the array before another is able to read the original
+value.  This is also called an [anti-dependency or WAR (write after
+read)](https://en.wikipedia.org/wiki/Data_dependency#Anti-dependency). In a
+functional language _nothing_ is in-place. Although it looks like `xs` is being
+updated in the example above, semantically we are creating a new array that just
+happens to have the same name. As a result, the reads and writes to `xs` must be
+separated, which Futhark does by transforming the code above into the following:
+
+```
+def stencil [n] (xs: *[n]i64): [n]i64 =
+  let tmp = map (\i -> xs[i-1] + xs[i] + xs[i+1]) (1..<n-1)
+  in xs with [1:n-1] = tmp
+```
+
+An auxilliary buffer is introduced and is used to guarantee that updates to `xs`
+do not interfere with any of the reads.  The problem is that this intermediate
+array can be unnecessary.  Consider this piece of code, which updates one row of
+a two-dimensional array:
+
+```
+def update [n][m] (i: i64) (xs: *[n][m]i64): [n][m]i64 =
+  xs with [i] = map (+1) xs[i]
+```
+
+As programmers, it is easy to determine that this update would be safe to do
+entirely in-place.  Each updated cell of the row depends only on the previous
+value of that same cell, and so they can be computed and written entirely in
+parallel with no risk of data-dependency errors.  Unfortunately, Futhark was not
+able to do so.  The conservative nature of high-level data-race guarantees, such
+as the ones Futhark provide, result in unnecessary memory being allocated and
+extraneous copies.
+
+While it may seem easy to get rid of the intermediate allocations in the example
+above. After all, there are no free values in the function being mapped. But
+more complex cases can make it hard for the compiler to verify that there are no
+data-dependencies.  For instance, in our [NW
+benchmark](https://github.com/diku-dk/futhark-benchmarks/blob/master/rodinia/nw/nw.fut),
```

Comment:
Make this a permanent link (click `y` while on the GitHub page).

Reply:
I'm a little hesitant, because that version actually _doesn't_ exhibit the pattern I'm talking about. That's why I wanted to investigate whether it made sense to merge diku-dk/futhark-benchmarks#32.

Reply:
Well, you can link directly to whatever implementation illustrates your point (maybe in the paper-specific benchmark repository?).  *Nobody* will care what is actually in the main futhark-benchmarks repository.

Reply:
Done.